tools/online_editor/src/properties_widget.ts

```diff
@@ -53,21 +67,68 @@ export class PropertiesWidget extends Widget {
         };
     }
 
+    dispose(): void {
+        super.dispose();
+    }
+
     set_language_client(client: BaseLanguageClient | null) {
-        this.#language_client = client ?? this.#language_client;
+        if (client != null) {
+            this.#language_client = client;
+            const work = client.middleware.handleWorkDoneProgress;
+            const new_work = (
+                token: ProgressToken,
+                params:
+                    | WorkDoneProgressBegin
+                    | WorkDoneProgressReport
+                    | WorkDoneProgressEnd,
+                next: HandleWorkDoneProgressSignature,
+            ) => {
+                if (params.kind === "begin") {
+                    this.data_stale(extract_uri(params.message || ""));
+                } else if (params.kind === "end") {
+                    this.data_valid(extract_uri(params.message || ""));
+                }
+                if (work != null) {
+                    work(token, params, next);
+                }
+            };
+            this.#language_client.middleware.handleWorkDoneProgress = new_work;
+        }
     }
 
-    position_changed(uri: LspURI, version: number, position: LspPosition) {
+    data_stale(uri: string) {
+        if (uri == this.#propertiesView.current_data_uri) {
```

Comment:
Instead of doing the uri check twice (here and in the other function), I think that it would be simpler to do it once at the call site. It will also reduce the indentation.

Reply:
Both functions perform the same `uri == this.#propertiesView.current_data_uri` check and indent their entire body if true (plus `data_valid` also checks if `this.#current_position` is true'ish). Instead, the two functions could take no parameters.

As I mentioned in the other comment: I think those URI check will only expand in the future, to cover multiple documents and (worst case) ranges, so I think centralising that is better than duplicating.